### PR TITLE
grouped-gemm MoE checkpoint reshape for zero-cost checkpoint

### PR DESCRIPTION
**Pull request by adam-xiaoyao**
#### Before submitting

- [ ] Lint code. If there are lint issues, please format the code first.

```shell
# Install and register `pre-commit` in the project folder
pip install pre-commit && pre-commit install

# Process previous code files separately
pre-commit run --file XXXX.py
```

- [ ] Add test cases into `tests` folder. If there are codecov issues, please add tests cases first.

### PR types

New features
### PR changes

Others
### Description

grouped-gemm MoE checkpoint reshape for zero-cost checkpoint

### CODE REVIEW

File: paddleformers/trainer/trainer.py [hunk 1]
```diff
@@ -1198,6 +1198,27 @@ def get_metadata_file_name(path):
             return metadata_files[0]
 
         model_sharded_state_dict = self.model.sharded_state_dict()
+        # Auto-enable MoE grouped-gemm expert reshape so a checkpoint saved
+        # under a different dispatcher layout (allgather N-D vs non-allgather
+        # 2-D) loads without any manual aoa_config. The neutral `reshape`
+        # statements are no-ops when the checkpoint already matches the current
+        # layout, so injecting them is always safe. Filtered to keys that
+        # actually exist in this model, and de-duplicated for idempotency.
+        if not self.args.load_from_hf and hasattr(self.model, "_gen_grouped_gemm_reshape_aoa_config"):
```
Comment (risemeup1111) [inline paddleformers/trainer/trainer.py]: P2 当前树里没有任何模型实现 `_gen_grouped_gemm_reshape_aoa_config`，所以这个 `hasattr(...)` 分支永远不会进。结果这里不会自动注入任何 reshape 规则，加载还是要靠外部手工 `aoa_config`。

File: paddleformers/trainer/trainer.py [hunk 2]
```diff
@@ -1431,9 +1452,9 @@ def _assign_master_weights_to_model(master_weights):
                             f"shape {master_weights[param.name].shape} to param "
                             f"{param.name} shape{param.shape}"
                         )
-                        assert (
-                            param.shape == master_weights[param.name].shape
-                        ), f"got {param.shape} vs {master_weights[param.name].shape}"
+                        assert math.prod(param.shape) == math.prod(master_weights[param.name].shape), (
```
Comment (risemeup1111) [inline paddleformers/trainer/trainer.py]: P1 这里把所有 bf16 master weight 的校验都放宽成只比 element count。这样只要 numel 一致，哪怕 shape/维度顺序错了也会被静默 reshape 进模型，`recover_params_from_master_weight()` 里其它 bf16 参数同样受影响。建议只对 grouped-gemm 命中的 key 放宽，其他参数保留精确 shape 检查。

**Review by risemeup1111 — CHANGES_REQUESTED** [review]
这次改动里有两处需要再收紧：一处是 bf16 master weight 的 shape 校验被放宽得过宽，另一处是自动注入 grouped-gemm reshape 的 hook 目前没有实际落点。

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>

**Comment by risemeup1111** [comment]
### Nyanpasu Review Board


| 序号 | 位置 | 优先级 | 状态 |
| - | - | - | - |
| 1 | [paddleformers/trainer/trainer.py#discussion_r3672160103](https://github.com/PaddlePaddle/PaddleFormers/pull/4812#discussion_r3672160103) | P1 | ✅ |
| 2 | [paddleformers/trainer/trainer.py#discussion_r3672160004](https://github.com/PaddlePaddle/PaddleFormers/pull/4812#discussion_r3672160004) | P2 | ✅ |
| 3 | [paddleformers/trainer/trainer.py#discussion_r3672917570](https://github.com/PaddlePaddle/PaddleFormers/pull/4812#discussion_r3672917570) | P1 | 🚧 |

<details>
<summary>点击展开 Review 规则</summary>
PR 评审规则：

- P0、P1 级别的评审必须提交新的 commit 进行修改；
- P2、P3 级别的评审可以通过评论进行修改。

状态说明：

| 已解决 | 无需修复 | 待修复 |
| - | - | - |
| ✅ | 🟡 | 🚧 |
</details>

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>

File: paddleformers/trainer/trainer.py (hunk 2 above)
Comment (adam-xiaoyao) [reply paddleformers/trainer/trainer.py]: OK, 我已经改了， 但我忘了再用另一个commit了。

File: paddleformers/trainer/trainer.py (hunk 1 above)
Comment (adam-xiaoyao) [reply paddleformers/trainer/trainer.py]: OK, 我已经改了， 但我忘了再用另一个commit了。

File: paddleformers/trainer/trainer.py [hunk 3]
```diff
@@ -1317,8 +1317,58 @@ def get_metadata_file_name(path):
...
+                    if len(canonical) < 3:
+                        continue
+                    flattened = (canonical[0] * canonical[1], math.prod(canonical[2:]))
+                    if {src_shape, dst_shape} != {canonical, flattened}:
+                        continue
+                    # AOA keys weights by the base model-state name (the optimizer
+                    # suffix is stripped internally), so the statement must use the
+                    # suffix-free struct name on both sides.
+                    seen.add(struct_name)
+                    statements.append(f"{struct_name} -> {struct_name}, reshape = '{list(canonical)}'")
```
Comment (risemeup1111) [inline paddleformers/trainer/trainer.py:1358]: P1 这里生成的 `reshape = ...` 目前会被 Paddle AOA parser 判成不支持的属性；`shape_propagation()` 只处理 `axis`、`permute`、`dtype`、`src_dtype`、`dst_dtype`，没有 `reshape` 分支，命中这里会直接报 `Unsupported attribute`。

**Review by risemeup1111 — CHANGES_REQUESTED** [review]
当前还有一处阻断问题：新增的 grouped-gemm AOA 语句使用了当前解析器不支持的 `reshape` 属性，会在加载时直接报错。请先补齐 AOA 端支持或换成现有语义后再提交。

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>

File: paddleformers/trainer/trainer.py (hunk 3 above)
Comment (adam-xiaoyao) [reply paddleformers/trainer/trainer.py:1358]: OK, 我会先等paddle的合入

Comment (risemeup1111) [reply paddleformers/trainer/trainer.py:1358]: 收到，不过这条仍然是阻断状态：当前 PR 里生成的 `reshape` 语句在现有 Paddle AOA 解析器里还不支持。等上游合入后再同步这版实现即可；在那之前，这项问题还不能算解决。

**Review by zjjlivein — APPROVED** [review]

**Review by ForFishes — APPROVED** [review]
LGTM
